### DISCUSSION TITLE
Integration with externally deployed models

### DISCUSSION THREAD

**Discussion by KiK0S** [mlflow/mlflow#12720] [Q&A] [answered]
Hello,

We at [Datagrok](https://datagrok.ai) are developing a platform for data processing and analysis with a focus on chem- and bio-informatics. As part of our toolkit, we provide users with predictive models, either created on our platform or external ones. We currently use MLFlow for integrating external models. Specifically, we create a Docker image for a model using `mlflow models build-docker` on our platform and then allow users to enrich their data with model predictions directly in the table views. [This video](https://www.youtube.com/watch?v=RS163zKe7s8) demonstrates the feature.

As we enhance the user experience with this feature, we aim to not only support model inference on our platform but also to use models deployed externally (in Sagemaker, Kubernetes, etc.). We are seeking advice on the optimal way to achieve this.

Here are a few questions we have:

- What is the REST API for the model servers? Our current features are built based on the [local inference server specification](https://mlflow.org/docs/latest/deployment/deploy-model-locally.html#inference-server-specification). Is it the same for other deployment methods?

- Is there a registry of deployments in the tracking server? Our current implementation fetches existing models from the tracking server, but we haven't found an API in the tracking server for listing deployments. There is an `mlflow deployments` command, but it won't be accessible for us, as we plan to connect to a remote MLFlow tracking server.

We appreciate any insights or recommendations you can provide on these topics. Thank you!

**Comment by B-Step62** [comment, marked as answer]
Hi, @KiK0S! It's great to hear about the integration.

> What is the REST API for the model servers? Our current features are built based on the [local inference server specification](https://mlflow.org/docs/latest/deployment/deploy-model-locally.html#inference-server-specification). Is it the same for other deployment methods?

Yes, this specification should be same across different deployment targets. You should be able to query the model in a same manner if it is an model deployed via MLflow.

> Is there a registry of deployments in the tracking server? Our current implementation fetches existing models from the tracking server, but we haven't found an API in the tracking server for listing deployments. There is an mlflow deployments command, but it won't be accessible for us, as we plan to connect to a remote MLFlow tracking server.

Unfortunately, MLflow does not provide a built-in API for listing deployments directly from the tracking server. You might need to implement a custom solution to track and manage your deployments externally. One approach would be using [model version tags](https://mlflow.org/docs/latest/model-registry.html#deploy-and-organize-models-with-aliases-and-tags) in the MLflow Model Registry to record deployment metadata as simple key-value tags for each model version, such as deployment time, platform, etc. 
If it doesn't satisfy your workload, you may need to maintain a separate registry or metadata store that logs deployment details. 

I hope this helps clarify things for you.

**Reply by KiK0S** [reply]
Hello, @B-Step62! Thank you for your answer. I think model tags will work for us in a foreseeable future.
